Source/WebCore/rendering/style/RenderStyle.cpp

```diff
@@ -2431,6 +2431,7 @@ const Color& RenderStyle::visitedLinkColor() const
 
 void RenderStyle::setColor(const Color& v)
 {
+    SET_VAR(m_inheritedData, visitedLinkColor, v);
     SET_VAR(m_inheritedData, color, v);
 }
```

Comment:
This doesn't seem correct. Visited and unvisited colors are resolved separately. Setting visited link color is handled by BuilderCustom::applyValueColor which has special handling for `currentcolor`, treating it the same as `inherit`. There is a discrepency here:

```        
if (builderState.applyPropertyToVisitedLinkStyle()) {
    auto color = builderState.colorFromPrimitiveValue(primitiveValue, ForVisitedLink::Yes);
    builderState.style().setVisitedLinkColor(color.resolveColor(builderState.parentStyle().visitedLinkColor()));
}
```
vs.
```
    if (builderState.applyPropertyToVisitedLinkStyle())
        builderState.style().setVisitedLinkColor(builderState.parentStyle().color());
```
(inheriting from visitedLinkColor vs color)

Reply:
I suppose the underlying issue is that we don't compute visited link color unless we are inside link

```
inline void Builder::applyCascadeProperty(const PropertyCascade::Property& property)
{
    SetForScope levelScope(m_state.m_currentProperty, &property);

    auto applyWithLinkMatch = [&](SelectorChecker::LinkMatchMask linkMatch) {
        if (property.cssValue[linkMatch]) {
            SetForScope scopedLinkMatchMutation(m_state.m_linkMatch, linkMatch);
            applyProperty(property.id, *property.cssValue[linkMatch], linkMatch);
        }
    };

    applyWithLinkMatch(SelectorChecker::MatchDefault);

    if (m_state.style().insideLink() == InsideLink::NotInside)
        return;

    applyWithLinkMatch(SelectorChecker::MatchLink);
    applyWithLinkMatch(SelectorChecker::MatchVisited);

    m_state.m_linkMatch = SelectorChecker::MatchDefault;
}
```

but also don't know we haven't computed it.

Reply:
Would something like

```
diff --git a/Source/WebCore/style/StyleBuilder.cpp b/Source/WebCore/style/StyleBuilder.cpp
index c725151eb7fc..f617496e0914 100644
--- a/Source/WebCore/style/StyleBuilder.cpp
+++ b/Source/WebCore/style/StyleBuilder.cpp
@@ -344,7 +344,7 @@ void Builder::applyProperty(CSSPropertyID id, CSSValue& value, SelectorChecker::
 
     ASSERT(!isInherit || !isInitial); // isInherit -> !isInitial && isInitial -> !isInherit
 
-    if (m_state.applyPropertyToVisitedLinkStyle() && !isValidVisitedLinkProperty(id)) {
+    if (!m_state.applyPropertyToRegularStyle() && !isValidVisitedLinkProperty(id)) {
         // Limit the properties that can be applied to only the ones honored by :visited.
         return;
     }
diff --git a/Source/WebCore/style/StyleBuilderState.h b/Source/WebCore/style/StyleBuilderState.h
index 891cd9c6c20c..0421fb33914a 100644
--- a/Source/WebCore/style/StyleBuilderState.h
+++ b/Source/WebCore/style/StyleBuilderState.h
@@ -85,7 +85,7 @@ public:
 
     // FIXME: These are mutually exclusive, clean up the code to take that into account.
     bool applyPropertyToRegularStyle() const { return m_linkMatch != SelectorChecker::MatchVisited; }
-    bool applyPropertyToVisitedLinkStyle() const { return m_linkMatch == SelectorChecker::MatchVisited; }
+    bool applyPropertyToVisitedLinkStyle() const { return m_linkMatch != SelectorChecker::MatchLink; }
 
     bool useSVGZoomRules() const;
     bool useSVGZoomRulesForLength() const;
```

work?

Reply:
Yes it's way cleaner. 👍  I'm running AB testing perf and memory because we are gonna compute the property value was more often now.